lib/exq/redis/job_queue.ex

```diff
@@ -106,17 +106,31 @@ defmodule Exq.Redis.JobQueue do
     end
   end
 
+  @reload_backup_queue_script Connection.prepare_script """
+    local jobs_to_reload = redis.call('LLEN', KEYS[1])
+    local backup_queue_reversed_key = KEYS[1] .. "::reversed"
+
+    while redis.call('LLEN', KEYS[1]) > 0 do
```

Comment:
Do we need the reverse queue?  Seems like we should be able to use RPOP or LPOP depending on order required. 

Also, with this setup, I think it's enqueueing the list in reverse, the oldest ones from backup queue:
```
# with concurrency 3  and sleep 15
iex(1)> Exq.enqueue(Exq, "default", Exq.MyWorker, [])
{:ok, "4ed66da6-c217-4f28-aeaf-40431bdb21c8"}
iex(2)>
2018-02-11 11:41:43.568 [info]: Elixir.Exq.MyWorker[4ed66da6-c217-4f28-aeaf-40431bdb21c8] start
iex(2)> Exq.enqueue(Exq, "default", Exq.MyWorker, [])
{:ok, "05b9cf87-da56-4371-b031-ae3e8c82d025"}
iex(3)>
2018-02-11 11:41:43.928 [info]: Elixir.Exq.MyWorker[05b9cf87-da56-4371-b031-ae3e8c82d025] start
iex(3)> Exq.enqueue(Exq, "default", Exq.MyWorker, [])
{:ok, "f830fb70-d16c-4d7b-9f5e-61ac6e7881cd"}
iex(4)>
2018-02-11 11:41:44.393 [info]: Elixir.Exq.MyWorker[f830fb70-d16c-4d7b-9f5e-61ac6e7881cd] start
iex(4)> Exq.enqueue(Exq, "default", Exq.MyWorker, [])
{:ok, "b9272ddc-693a-403e-953e-c9934403fa96"}
iex(5)> Exq.enqueue(Exq, "default", Exq.MyWorker, [])
{:ok, "ac833ca5-34b9-4e74-b1cd-03b1fb30cb6c"}
iex(6)> Exq.enqueue(Exq, "default", Exq.MyWorker, [])
{:ok, "af78085d-f54e-4f50-8b0c-e6c8c3d46280"}
iex(7)> Exq.enqueue(Exq, "default", Exq.MyWorker, [])
{:ok, "19539a40-4569-4d10-ac35-8c7fa9c4e746"}
iex(8)>
```

Killing and restarting this:
```
# with concurrency 1 and sleep 15
2018-02-11 11:41:58.163 [info]: Re-enqueued 3 job(s) from backup for node_id [MacBook-Pro] and queue [default]
iex(1)>
2018-02-11 11:41:58.171 [info]: Elixir.Exq.MyWorker[f830fb70-d16c-4d7b-9f5e-61ac6e7881cd] start
2018-02-11 11:42:13.178 [info]: Elixir.Exq.MyWorker[f830fb70-d16c-4d7b-9f5e-61ac6e7881cd] done: 15007ms sec
2018-02-11 11:42:13.179 [info]: Elixir.Exq.MyWorker[05b9cf87-da56-4371-b031-ae3e8c82d025] start
2018-02-11 11:42:28.181 [info]: Elixir.Exq.MyWorker[05b9cf87-da56-4371-b031-ae3e8c82d025] done: 15001ms sec
2018-02-11 11:42:28.182 [info]: Elixir.Exq.MyWorker[4ed66da6-c217-4f28-aeaf-40431bdb21c8] start
2018-02-11 11:42:43.185 [info]: Elixir.Exq.MyWorker[4ed66da6-c217-4f28-aeaf-40431bdb21c8] done: 15003ms sec
2018-02-11 11:42:43.186 [info]: Elixir.Exq.MyWorker[b9272ddc-693a-403e-953e-c9934403fa96] start
2018-02-11 11:42:58.188 [info]: Elixir.Exq.MyWorker[b9272ddc-693a-403e-953e-c9934403fa96] done: 15001ms sec
2018-02-11 11:42:58.189 [info]: Elixir.Exq.MyWorker[ac833ca5-34b9-4e74-b1cd-03b1fb30cb6c] start
2018-02-11 11:43:13.194 [info]: Elixir.Exq.MyWorker[ac833ca5-34b9-4e74-b1cd-03b1fb30cb6c] done: 15005ms sec
2018-02-11 11:43:13.195 [info]: Elixir.Exq.MyWorker[af78085d-f54e-4f50-8b0c-e6c8c3d46280] start
2018-02-11 11:43:28.198 [info]: Elixir.Exq.MyWorker[af78085d-f54e-4f50-8b0c-e6c8c3d46280] done: 15002ms sec
2018-02-11 11:43:28.198 [info]: Elixir.Exq.MyWorker[19539a40-4569-4d10-ac35-8c7fa9c4e746] start
2018-02-11 11:43:43.201 [info]: Elixir.Exq.MyWorker[19539a40-4569-4d10-ac35-8c7fa9c4e746] done: 15002ms sec
```

Reply:
The reverse queue is necessary to ensure that the jobs are, as you phrase it below, "re-enqueued in the same order as originally queued." If the jobs were popped off the backup queue and pushed onto the job queue one-by-one, they would end up in reverse order. The temporary queue is the low-memory-usage way to move the items onto the queue in the "correct" order.

(For an alternative, you could query the entire backup queue into a Lua table, and then iterate through it in reverse order, pushing the items into the job queue. This would require that the Redis instance could hold the backup queue in memory, which may or may not be reasonable, and—from my understanding of Lua table iteration—would be a fair bit slower.)

Reply:
Maybe I'm missing something - Is there a reason we can’t just pop from the other side of the queue using reverse command LPOP or RPOP in the Lua script?  Also the logic was wrong when I tested this locally, it used the wrong order as it enqueued newer ones before the older messages for the ones in the backup - see output from above.